File: src/apps/MultitouchExtension/src/MultitouchDeviceManager.m
#import "MultitouchDeviceManager.h"
#import "FingerStatusManager.h"
#import "KarabinerKit/KarabinerKit.h"

//
// Multitouch callback
//

static int callback(MTDeviceRef device,
                    Finger* data,
                    int fingers,
                    double timestamp,
                    int frame) {
  if (!data) {
    fingers = 0;
  }

  if (device) {
    FingerStatusManager* manager = [FingerStatusManager sharedFingerStatusManager];
    [manager update:device
               data:data
            fingers:fingers
          timestamp:timestamp
              frame:frame];
  }

  return 0;
}

@interface MultitouchDeviceManager ()

@property(copy) NSArray* devices;
@property IONotificationPortRef notificationPort;
@property CFRunLoopSourceRef loopSource;

@end

@implementation MultitouchDeviceManager

- (instancetype)init {
  self = [super init];

  if (self) {
    _devices = [NSMutableArray new];
  }

  return self;
}

+ (instancetype)sharedMultitouchDeviceManager {
  static dispatch_once_t once;
  static MultitouchDeviceManager* manager;

  dispatch_once(&once, ^{
    manager = [MultitouchDeviceManager new];
  });

  return manager;
}

- (void)setCallback:(BOOL)set {
  @synchronized(self) {
    //
    // Unset callback (even if set is YES.)
    //

    if (self.devices) {
      for (NSUInteger i = 0; i < [self.devices count]; ++i) {
        MTDeviceRef device = (__bridge MTDeviceRef)(self.devices[i]);
        if (!device) continue;

        MTDeviceStop(device, 0);
        MTUnregisterContactFrameCallback(device, callback);
      }

      self.devices = [NSArray new];
    }

    //
    // Set callback if needed
    //

    if (set) {
      self.devices = (NSArray*)CFBridgingRelease(MTDeviceCreateList());
      if (self.devices) {
        for (NSUInteger i = 0; i < [self.devices count]; ++i) {
          MTDeviceRef device = (__bridge MTDeviceRef)(self.devices[i]);
          if (!device) continue;

          MTRegisterContactFrameCallback(device, callback);
          MTDeviceStart(device, 0);
        }
      }
    }
  }
}

//
// IONotification
//

- (void)releaseIterator:(io_iterator_t)iterator {
  for (;;) {
    io_object_t obj = IOIteratorNext(iterator);
    if (!obj) break;

    IOObjectRelease(obj);
  }
}

static void relaunch(void* refcon, io_iterator_t iterator) {
  // Relaunch when devices are plugged/unplugged.
  [KarabinerKit relaunch];
}

- (void)registerIONotification {
  NSLog(@"registerIONotification");

  @synchronized(self) {
    if (self.notificationPort) {
      [self unregisterIONotification];
    }

    self.notificationPort = IONotificationPortCreate(kIOMasterPortDefault);
    if (!self.notificationPort) {
      NSLog(@"[ERROR] IONotificationPortCreate");
      return;
    }

    {
      // ------------------------------------------------------------
      NSMutableDictionary* match = (__bridge NSMutableDictionary*)(IOServiceMatching("AppleMultitouchDevice"));

      // ----------------------------------------------------------------------
      io_iterator_t it;
      kern_return_t kr;

      // for kIOTerminatedNotification
      kr = IOServiceAddMatchingNotification(self.notificationPort,
                                            kIOTerminatedNotification,
                                            (__bridge CFMutableDictionaryRef)(match),
                                            relaunch,
                                            (__bridge void*)(self),
                                            &it);
      if (kr != kIOReturnSuccess) {
        NSLog(@"[ERROR] IOServiceAddMatchingNotification");
        return;
      }
      [self releaseIterator:it];
    }

    {
      // ------------------------------------------------------------
      NSMutableDictionary* match = (__bridge NSMutableDictionary*)(IOServiceMatching("AppleMultitouchDevice"));

      // ----------------------------------------------------------------------
      io_iterator_t it;
      kern_return_t kr;

      // for kIOMatchedNotification
      kr = IOServiceAddMatchingNotification(self.notificationPort,
                                            kIOMatchedNotification,
                                            (__bridge CFMutableDictionaryRef)(match),
                                            relaunch,
                                            (__bridge void*)(self),
                                            &it);
      if (kr != kIOReturnSuccess) {
        NSLog(@"[ERROR] IOServiceAddMatchingNotification");
        return;
      }
      [self releaseIterator:it];
    }

    // ----------------------------------------------------------------------
    self.loopSource = IONotificationPortGetRunLoopSource(self.notificationPort);
    if (!self.loopSource) {
      NSLog(@"[ERROR] IONotificationPortGetRunLoopSource");
      return;
    }
    CFRunLoopAddSource(CFRunLoopGetCurrent(), self.loopSource, kCFRunLoopDefaultMode);
  }
}

- (void)unregisterIONotification {
  NSLog(@"unregisterIONotification");

  @synchronized(self) {
    if (self.notificationPort) {
      if (self.loopSource) {
        CFRunLoopSourceInvalidate(self.loopSource);
        self.loopSource = nil;
      }
      IONotificationPortDestroy(self.notificationPort);
      self.notificationPort = nil;
    }
  }
}

@end

File: src/apps/MultitouchExtension/src/MultitouchDeviceManager.swift
func mtDeviceRegistryEntryID(of device: MTDevice?) -> UInt64? {
  let service = MTDeviceGetService(device)
  var entryID: UInt64 = 0
  let kr = IORegistryEntryGetRegistryEntryID(service, &entryID)

  guard kr == KERN_SUCCESS else {
    return nil
  }

  return entryID
}

private func callback(
  _ device: MTDevice?,
  _ fingerData: UnsafeMutablePointer<Finger>?,
  _ fingerCount: Int32,
  _ timestamp: Double,
  _ frame: Int32
) -> Int32 {
  let fingers =
    fingerData != nil
    ? Array(UnsafeBufferPointer(start: fingerData, count: Int(fingerCount)))
    : []

  if let device = device {
    if let mtDeviceRegistryEntryID = mtDeviceRegistryEntryID(of: device) {
      Task { @MainActor in
        FingerManager.shared.update(
          mtDeviceRegistryEntryID: mtDeviceRegistryEntryID,
          fingers: fingers,
          timestamp: timestamp,
          frame: frame)
      }
    }
  }

  return 0
}

@MainActor
class MultitouchDeviceManager {
  static let shared = MultitouchDeviceManager()

  private var notificationsTask: Task<Void, Never>?

  private let notificationPort = IONotificationPortCreate(kIOMainPortDefault)

  private var devices: [MTDevice] = []
  private var wakeObserver: NSObjectProtocol?

  init() {
    notificationsTask = Task {
      await withTaskGroup(of: Void.self) { group in
        group.addTask {
          for await _ in NSWorkspace.shared.notificationCenter.notifications(
            named: NSWorkspace.didWakeNotification
          ) {
            Task { @MainActor in
              print("didWakeNotification")

              do {
                // Sleep until devices are settled.
                try await Task.sleep(nanoseconds: NSEC_PER_SEC)

                if UserSettings.shared.relaunchAfterWakeUpFromSleep {
                  try await Task.sleep(
                    nanoseconds: UInt64(UserSettings.shared.relaunchWait) * NSEC_PER_SEC)

                  // MultitouchExtension will be relaunched by launchd.
                  NSApplication.shared.terminate(self)
                }

                MultitouchDeviceManager.shared.setCallback(true)
              } catch {
                print(error.localizedDescription)
              }
            }
          }
        }
      }
    }
  }

  func setCallback(_ register: Bool) {
    print("setCallback \(register)")

    //
    // Unset callback first
    //

    for device in devices {
      MTDeviceStop(device, 0)
      MTUnregisterContactFrameCallback(device, callback)
    }

    devices.removeAll()

    //
    // Set callbacodk
    //

    if register {
      devices = (MTDeviceCreateList().takeUnretainedValue() as? [MTDevice]) ?? []

      for device in devices {
        MTRegisterContactFrameCallback(device, callback)
        MTDeviceStart(device, 0)
      }
    }
  }

  //
  // IONotification
  //

  private func releaseIterator(_ iterator: io_iterator_t) {
    while true {
      let obj = IOIteratorNext(iterator)
      if obj == 0 {
        break
      }

      IOObjectRelease(obj)
    }
  }

  func observeIONotification() {
    print("observeIONotification")

    //
    // Relaunch if device is connected or disconnected
    //

    let match = IOServiceMatching("AppleMultitouchDevice") as NSMutableDictionary

    for notification in [
      kIOMatchedNotification,
      kIOTerminatedNotification,
    ] {
      do {
        var it: io_iterator_t = 0

        let kr = IOServiceAddMatchingNotification(
          notificationPort,
          notification,
          match,
          { _, _ in
            // MultitouchExtension will be relaunched by launchd.
            NSApplication.shared.terminate(nil)
          },
          nil,
          &it)
        if kr != kIOReturnSuccess {
          print("IOServiceAddMatchingNotification error: \(kr)")
          return
        }

        releaseIterator(it)
      }
    }

    let loopSource = IONotificationPortGetRunLoopSource(notificationPort).takeUnretainedValue()
    CFRunLoopAddSource(RunLoop.current.getCFRunLoop(), loopSource, .defaultMode)
  }
}
